src/QsCompiler/CompilerOptimizations/SideEffectChecking.fs

```diff
@@ -8,24 +8,36 @@ open Microsoft.Quantum.QsCompiler.Transformations.Core
 open Utils
 
 
+/// A ScopeTransformation that tracks what side effects the transformed code could cause
 type internal SideEffectChecker() =
     inherit ScopeTransformation()
 
     let mutable anyQuantum = false
     let mutable anyMutation = false
     let mutable anyInterrupts = false
+    let mutable anyOutput = false
 
+    /// Whether the transformed code might have any quantum side effects (such as calling operations)
     member __.hasQuantum = anyQuantum
+    /// Whether the transformed code might change the value of any mutable variable
     member __.hasMutation = anyMutation
+    /// Whether the transformed code has any statements that interrupt normal control flow (such as returns)
     member __.hasInterrupts = anyInterrupts
+    /// Whether the transformed code might output any messages to the console
+    member __.hasOutput = anyOutput
 
     override __.Expression = { new ExpressionTransformation() with
         override expr.Kind = { new ExpressionKindTransformation() with
             override __.ExpressionTransformation ex = expr.Transform ex
             override __.TypeTransformation t = expr.Type.Transform t
 
+            override __.onFunctionCall (method, arg) =
+                anyOutput <- true
```

Comment:
It seems you are indicating the transformation might output messages if there is any call whatsoever - that seems rather pessimistic. Why is that?

Reply:
I don't know for sure what other functions a given function might call, so I make the pessimistic assumption that it might eventually call a Message function. While this means that I'll almost never remove useless function calls, this doesn't seem like a very important optimization, as in almost all cases it will only affect classical performance. It also means that Message functions aren't removed, which is desirable in most cases. It'd be good to have a compiler option later that determines whether we should remove all functions anyways.

Reply:
Hmm, it would be good to have a handle to enable or disable this special accomodation for Message. I don't remember - is there such a handle?

Reply:
No, I don't currently have a handle like that. Where would be the best place to put one?

Reply:
You would need to have the handle on the transformation that eliminates the messages, and propagate it up to the front handle that defines which of transformations to run in which order. The behavior of all other transformations should be independent on that configuration. Is that the case?

Reply:
I added the handle in Optimizations.fs, the file that defines which transformations to run in which order.

Reply:
I added the handle in Optimizations.fs, the file that defines which transformations to run in which order.